client_test.go

```diff
@@ -3,72 +3,63 @@ package zooz
 import (
 	"bytes"
 	"context"
-	"encoding/json"
 	"io/ioutil"
 	"net/http"
-	"reflect"
 	"testing"
 
 	"github.com/pkg/errors"
+	"github.com/stretchr/testify/require"
 )
 
 type httpClientMock struct {
-	do func(r *http.Request) (*http.Response, error)
-}
+	t *testing.T
 
-type request struct {
-	Field string `json:"field"`
-}
+	expectedMethod   string
+	expectedURL      string // can be relative i.e. "/somepath?k=v", in this case will be prepended with ApiURL
+	expectedHeaders  map[string]string
+	expectedBodyJSON string
 
-type callerMock struct {
-	t               *testing.T
-	expectedMethod  string
-	expectedPath    string
-	expectedHeaders map[string]string
-	expectedReqObj  interface{}
-	returnRespObj   interface{}
-	returnError     error
+	responseCode int
+	responseBody string
+	error        error
 }
 
 func (c *httpClientMock) Do(r *http.Request) (*http.Response, error) {
-	return c.do(r)
-}
-
-func (c *callerMock) Call(ctx context.Context, method, path string, headers map[string]string, reqObj interface{}, respObj interface{}) error {
-	if method != c.expectedMethod {
-		c.t.Errorf("Invalid method: %s", method)
-	}
-	if path != c.expectedPath {
-		c.t.Errorf("Invalid path: %s", path)
-	}
-	for k, v := range headers {
-		if v != c.expectedHeaders[k] {
-			c.t.Errorf("Invalid header %s: %s", k, v)
-		}
-	}
-	if len(headers) != len(c.expectedHeaders) {
-		c.t.Errorf("Invalid headers count: %d", len(headers))
+	require.Equal(c.t, c.expectedMethod, r.Method)
+	if len(c.expectedURL) > 0 && c.expectedURL[0] == '/' {
+		c.expectedURL = ApiURL + c.expectedURL
 	}
+	require.Equal(c.t, c.expectedURL, r.URL.String())
 
-	reqBody, err := json.Marshal(reqObj)
-	if err != nil {
-		c.t.Fatalf("Marshal error: %s", err)
+	for k, v := range c.expectedHeaders {
+		require.Equal(c.t, v, r.Header.Get(k))
```

Comment:
Now you assert that expected headers is a subset of actual headers. Don't you want to assert that they are equal?

Reply:
Generally, we are interested in endpoint-specific special headers, such as idempotency key, client user agent, client ip address).
Copy-pasting full list of headers would be reduntant.
Adding them as defaults in httpClientMock is an option, but i decided that testing a full list of headers once in `TestCall_WithApiResponse` is enough.

Reply:
Added more headers for assertion in `TestCall_WithApiResponse`, including content type and api version,